### PR TITLE
Feature - add mining fallback for idling warehouse fleet miner ships

### PR DESCRIPTION
**Pull request by VasiliyTemniy**
# PR: Idle-mining fallback for mining ships ("Mine When Idle")

## Motivation

Solid/liquid wares (ore, silicon, ice, hydrogen, methane, helium) can only be transported by mining ships, so warehouse networks that distribute raw resources are necessarily crewed by miners. Those ships idle a lot when the network is saturated — and unlike container traders they can do something genuinely useful while idle: mine the very wares their home warehouse stocks.

This PR adds an opt-in, per-fleet "Mine When Idle" option. When the scheduler repeatedly finds no trade for a mining-capable ship, the ship flies out, mines the home warehouse's most under-stocked mineable ware, and delivers it back through normal trade orders.

## What it does

1. **Trigger**: when `WarehouseFleetScheduler` ends a cycle with "No trade found" for a ship whose fleet has `miningFallback` enabled and which has solid or liquid cargo capacity, a per-ship idle clock starts. Only after the ship has stayed jobless for a configurable time (global setting, default 60s) is a mining trip dispatched. The clock is cleared whenever the ship gets real work or has any active order, so mining never fires in the middle of a delivery.
2. **Ware selection** (`PickMiningWare`): among ore/silicon/ice (solid) and hydrogen/methane/helium (liquid — nividium is deliberately excluded), pick the ware with the lowest `cargo/target` ratio at the home warehouse, but only if the warehouse has a buy offer for it, a storage allocation, and projected room left after accounting for **in-flight mining reservations** — concurrent trips register in a shared `$miningOps` table, so multiple miners never all fetch the same ware for the same half-empty tank. Stale reservations (ship died mid-trip, behaviour changed) are aged out by the existing GarbageCollector cue after 35 minutes, mirroring the 30-minute delivery assumption already used there.
3. **Mining trip** (`aiscripts/order.warehousefleet.mining.xml`, new): a one-shot internal order. Vanilla `MiningRoutine` is an infinite mine→sell loop and fights the mod's own delivery flow, so this order instead scores candidate sectors around the home sector by `yieldrating / (1 + gateDistance * 0.4)` (respecting sector blacklists via `lib.find.sectors.inrange`), picks a resource patch with a randomized min-rating floor (spreads ships across patches instead of converging on the single best rock), dispatches a single vanilla `MiningCollect` sub-order, releases the yield reservation, and ends. The ship then falls back to its WarehouseFleet default behaviour.
4. **Delivery** (`DeliverMinedCargoOrFallback`): on the next scheduler cycle the cargo aboard is routed: Supply Targets first, then the home warehouse, via real trade orders that participate in the mod's delivery/reservation bookkeeping; only wares that neither can take fall through to the existing `EmptyCargoRoutine` (`failedOrderHandling`). Afterwards a `ReturnToWarehouse` order is queued so follow-up mining starts from the correct anchor.

## New settings

- Fleet parameter **Mine When Idle** (`miningFallback`, bool, **default off** — existing fleets are unaffected; `<patch sinceversion="44">` heals running order instances in existing saves).
- Fleet parameter **Mining: Max Gate Distance** (`miningMaxGateDistance`, 0–10, default 3): how far from the home sector the trip may search; 0 = home sector only.
- Global setting **Mining fallback idle time** (slider in the existing Extension Options menu, default 60s), seeded via `Config` cue `<patch sinceversion="8">`.

## Behavior change for mining-capable ships with leftover cargo

Previously, any leftover cargo at scheduler entry went straight to `EmptyCargoRoutine` (refund/teleport/destroy/return/drop). For ships with solid/liquid capacity this PR routes leftover cargo through the supply-target → home-warehouse delivery path first, which is strictly gentler; `EmptyCargoRoutine` remains the final fallback. **Container-only ships are completely unaffected** — they keep the original path.

## Files changed

| File | Change |
| --- | --- |
| `aiscripts/order.warehousefleet.mining.xml` | **new** — one-shot mining order (`WarehouseFleetMine`) |
| `aiscripts/order.warehousefleet.xml` | version 43→44, two new params with `sinceversion="44"` patches |
| `md/warehousefleets.xml` | Config cue v7→8 (+ idle-time global); main cue v15→16 (+ `$miningOps`/`$idleSince`); scheduler hook; `TryMiningFallback`, `PickMiningWare`, `DeliverMinedCargoOrFallback` libraries; GarbageCollector cleanup |
| `md/warehousefleets_menu.xml` | one slider row (global settings section) |
| `libraries/icons.xml` | icon for the new order (uses the current `assets\textures\ui\order\` path — the pre-9.0 `assets\fx\gui\textures\` path used by the two existing entries no longer resolves in 9.00, left untouched here to keep this PR focused) |
| `readme.md` | documents the new fleet parameters, global setting, and a short feature section |

## Save compatibility

- No existing blocking actions in any aiscript are touched (the two new params are metadata only), so suspended order states import cleanly.
- New order params are patched into running `WarehouseFleet` order instances via `sinceversion="44"`.
- The main cue version bump (15→16) triggers the mod's existing reset path, which initializes the new bookkeeping tables on existing saves.
- Everything is opt-in; with "Mine When Idle" left off, the only behavior difference is the leftover-cargo routing described above.

## Notes for review

- One deliberate design point: `WarehouseFleetMine` requires the ship to be idle in the strict sense (`$ship.orders.count == 0`) before the idle clock even ticks, because `ScheduleFollowUpTripWhenDocking` pre-plans trips while the previous trade is still executing.
- The new aiscript declares a `debugchance` param (default 0) even though its own file logging uses the mod's usual always-on `debug_to_file` style: the vanilla interrupt handlers it references (e.g. `interrupt.inspected`) read `$debugchance` from the interrupted script's scope and raise "Property lookup failed" if it is undefined. (The same applies to the existing `WarehouseFleet` order — that fix is already committed in another PR.)

### CODE REVIEW

**Comment by mbleichner** [comment]
Tbh I'm a bit split on this. I can understand the wish to add such a functionality, but I see a few problems in its current form:

- Instead of time-based decision-making, there should be a priority setting that determines if a mining ship should do a transport or a mining job (just as all the other settings in the mod)
- After mining, the ship should deliver to the warehouse, not directly to a factory (the idea being that every ware flow always involves the warehouse)
- The game already has a mining routine, which should be used instead of the custom order (you're kind of reinventing the wheel here; also the comment in the file is wrong btw - you want the normal "Mine" order, not AutoMine)
- Depending on how exhausted your resource fields are, mining can take a loooong time - effectively deadlocking ore transport (~20 minutes in my case)

The goal of my mod was not to reimplement every part of the game, but to solve one thing well - logistics. Mining is kind of a different thing, imo.

As a side note: In the first version of the mod, transporting raw materials wasn't even possible, since I hadn't even considered the possibility anyone would want to do that.
I just add some miners to my factories and don't even bother with transporting raw materials.

**Comment by VasiliyTemniy** [comment]
If you would only know how much time I've spent trying to reuse vanilla mine order :)

It just does not fit.

The "Mine" order produces some visual sphere around resources - it's totally not what I wanted. It worked, but in a really weird way and it required the same setup of finding resources and such - it does not handle finding resources.
The "AutoMine" is actually the thing that I wanted to use. Why? Because it finds resources on its own, based on parameters, crew experience and such. But - it really does the exact thing stated in the huge ai-generated sloppy comment, right. Exactly that - "is an infinite loop (mine -> sell -> mine -> ...)" - I've tried to fix that, but required injections to vanilla aiscript. I could, obviously, do injections. But I think you would like that a lot less.

So - I came up with "do a minimal copy of auto mine" - exactly auto mine, that works as I wanted - no visual gather zone sphere, so on. So - a minimal copy.

I personally think it's a lot less harmful than any injection. Neither Mine nor AutoMine worked out for me.

----------------------------

Coming back to your comment - well, you're not obliged to accept it - at all :).

"As a side note: In the first version of the mod, transporting raw materials wasn't even possible, since I hadn't even considered the possibility anyone would want to do that." - in my case, it's the only thing I use it for - distribution is good, but I have a pipeline for produced wares, I only ever need it for the raw resources, and only in the "supply" mode. So - I've seen lots of my ships stalling doing nothing at some point. And I've decided to try to fill that gap.

"I just add some miners to my factories and don't even bother with transporting raw materials." - I've done the same in my previous playthtough, but this approach with warehouses and mining ships works like a charm.

One big warehouse used exclusively for raw materials storage, all miners assigned to that warehouse with warehouse fleet order - they gather resources and deliver to stations in need

**Comment by mbleichner** [comment]
> The "Mine" order produces some visual sphere around resources - it's totally not what I wanted. It worked, but in a really weird way and it required the same setup of finding resources and such - it does not handle finding resources.
> The "AutoMine" is actually the thing that I wanted to use. Why? Because it finds resources on its own, based on parameters, crew experience and such. But - it really does the exact thing stated in the huge ai-generated sloppy comment, right. Exactly that - "is an infinite loop (mine -> sell -> mine -> ...)" - I've tried to fix that, but required injections to vanilla aiscript. I could, obviously, do injections. But I think you would like that a lot less.

Yeah I haven't thought of the sphere thing. Sounds like a nightmare to determine the "correct" spot to send the ship to.

Personally, I would separate those two concerns, though - have a fleet dedicated to mining and one dedicated to transporting. Less efficient, but way more predictable and resilient to deadlocks, imo.

**Comment by VasiliyTemniy** [comment]
It's ok - but having separated "miners" and "mules" (transport job) would make "miners" semantically outside of warehouse fleet mod.

I will keep it in my version - and I will not post it anywhere. I can delete that branch in my fork, if you wish.

I had strong suspicions that you wont like this specific PR :). But the base of WarehouseFleets is just too good and creating another mod for miners solely would have to reimplement all the same, but with some weird cuts - and the result would be ... just bad. The vanilla's AutoMine does cover everything in that direction.

So - you can close this one, or I will do it, I don't think there's a possibility to stitch two things together otherwise.
Well, anyway - you are free to do whatever with this branch, if you ever feel like it - take anything, I've tested that code enough to assure you it's working good :)

**Comment by mbleichner** [comment]
Tbh I don't hate the idea of using idle ships for automining. It actually kind of is the only missing part of the puzzle to have a complete economy driven by the mod.

To reiterate on the problems that I see:
- The timing-based scheduling effectively prevents you from saying "mining is more important than transporting raw materials". Even if there is super abundant ore nearby. I think a priority-based system would be better here. The priorities need to be balanced against the existing ones, though.
- Delivering directly to a consumer station has the drawback that ore cannot be distributed to a neighbor warehouse, even if it is sorely needed there. So I would always deliver to the local warehouse first, so that a better decision can be made in a second round of scheduling.
- What worries me most is the danger of locking up a mining ship (or a whole fleet of mining ship) for too long if resources are scarce. Maybe this can be mitigated by adding some kind of timer to the mining behavior to cancel orders that are taking exceedingly long.

**Comment by VasiliyTemniy** [comment]
Ok!

**On priorities:**
Agreed, priority-based is a better fit for the mod's spirit. But I don't want a plain prioMine slider that makes miners mine while the warehouse is already full of ore - that's just a miner in disguise :) (but the slider will still be present - for the "balanced" decision-making threshold from below). My current idea:

- warehouse has more than min(3 full ship cargo bays, ware storage capacity) of some mineable ware -> mining effectively never competes, it fires only when no other job exists (almost a pure fallback)
- less than that -> mining enters the same importance/rating competition as distribution/supply/trade

Probably something exponential on the stock level: effectively zero importance above that threshold, balanced importance around ~3 ship loads, big importance when the warehouse is really starving. The timer goes away completely. I think some solid approach will come directly during development.
In other words, my intent is to make mining come into play when there's a real need for raw resources.

**On delivering to warehouse first:**
Agreed, will do. It's also the smallest change of the three - the whole supply-target delivery pass just gets deleted, mined cargo is sold to the home warehouse and the next scheduling round decides where it goes with full priority machinery.

**On the lockup:**
I have an md watcher in mind. Will come up with a heuristic based on time spent, mining "healthiness" (how much cargo gained since last check), distance from home warehouse, maybe an escalation counter - no heavy compute involved. Ship that mines well keeps mining, ship that has already mined something (arrived at the mining spot) and then scratches an exhausted field for a few minutes gets recalled with whatever it has.
Mechanically - a watcher rooted in WarehouseFleets cue, tracks stats per mining operation (per ship), makes a decision - to cancel or not to cancel an order.

I will try to reimplement it that way and come back :)

**Comment by mbleichner** [comment]
> warehouse has more than min(3 full ship cargo bays, ware storage capacity) of some mineable ware -> mining effectively never competes, it fires only when no other job exists (almost a pure fallback)

I don't see why such arbitrary thresholds are necessary. The goal should be to create a system that naturally self-regulates by weighing the different options against each other:

- both mining and distribution are candidates to resolve a low-storage situation
- **distribution importance** is proportional to the imbalance of the two involved stations (thinking of two warehouses here)
- **mining importance** should be inversely proportional to the local warehouse storage level
- which one of both should win is not clear and involves some kind of player choice/preference
- that's why **importance values are weighted by priorities**, so the player can influence how attractive a certain decision is
- the same principle should be applied here too:
  - `prioMining = prioDistribution` should make mining roughly equally attractive as fetching the resources from another warehouse
  - `prioMining < prioDistribution` means distribution is preferred over mining and and ships will only go mining if they can't find any significant imbalances to fix
  - `prioMining > prioDistribution` means mining is preferred over distribution; so unless there is a huge imbalance between stations, ships will rather mine than fix the imbalance
  - `prioMining = 0` always reduces importance to 0 (effectively disables mining; so we don't need an extra switch for that)
- Ideally, this all translates to some kind of formula - and we only need to look at the final importance values to decide on what to do (https://github.com/mbleichner/x4-warehouse-fleets/blob/master/md/warehousefleets.xml#L671-L684)

I hope I could clarify the idea...

**Comment by VasiliyTemniy** [comment]
You said literally what I've said, where "balance mining in such a way that it's most involved when system has low accumulated X ware" is my proposition for a "formula"

The only difference is "that's why importance values are weighted by priorities" - which I've just highlighted as "it should not be like if player has set all prios to 1 but mining to 10 then all ships just mine till the home warehouse is full".
Why should not? One reason only - all four existing "transportType" are moving ware X from A to B. There was "trade with allied npc stations if player wants it", but I read some of your comments somewhere that it's like not the topmost priority. There is _nothing_ yet to bring additional X to the warehouse system from outside of warehouse system. There is no "buy ware X from npc station C". So - (except for 'trade') all existing "transportType" are moving wares, not producing, not adding them to the system. Mining, on the other hand, is something that would bring additional resources to the system. This is the sole reason of why I wanted to make some difference - because if some ware X is abundant somewhere in the system already, it's faster to go take it and bring to the reciever than to mine it. That's it. It's the "base for my formula".

In other words - I understand what you say. I've agreed to all three of your previous points to give this idea a chance. When I say "I agree" - it does not mean that "you are right", it means "ok, whatever - it can also work even if I'd prefer the other way".

So - if you have a better formula than some formula that I did not have a chance yet to write more than in plain English - then why not?

"https://github.com/mbleichner/x4-warehouse-fleets/blob/master/md/warehousefleets.xml#L671-L684" - I knew where importance is calculated in this mod :). "Probably something exponential on the stock level: effectively zero importance above that threshold, balanced importance around ~3 ship loads, big importance when the warehouse is really starving. " from my previous message was literally planned to be placed near those lines. You said "balance" - it was my idea for "balance" for this particular activity

**Comment by mbleichner** [comment]
> "it should not be like if player has set all prios to 1 but mining to 10 then all ships just mine till the home warehouse is full"

Why not? seems like that's exactly what I wanted to achieve by setting the priorities like that? Of course, when the warehouse is full, the ships can't bring in full ore loads anymore, so the importance rating falls off a cliff, making distribution jobs more attractive (even if their prio is low).

> When I say "I agree" - it does not mean that "you are right", it means "ok, whatever - it can also work even if I'd prefer the other way".

I'll gladly discuss this with you. I'm not interested in "winning" an argument, but finding the best solution.

The reason I'm against hardcoded thresholds is that it's basically impossible to make them work reliably in every possible case. Some players might have really tiny storages that won't even fit 3 shiploads - others might have storages that are only 0.5% full with 3 shiploads.
Using relative thresholds instead (like 30% of storage capacity) is generally better, but can also lead to suboptimal situations depending on your fleet, your ore environment, distances and whatnot.

The only thing that always works in every scenario is weighing all the different options against each other, and choosing the one with the most efficient outcome. That's the whole idea I built the mod on.

In my experience, adding thresholds and conditions to such algorithms always starts simple - but becomes full-on hell later, as more and more such "discontinuities" get added in.

**Comment by VasiliyTemniy** [comment]
> The reason I'm against hardcoded thresholds is that it's basically impossible to make them work reliably in every possible case. Some players might have really tiny storages that won't even fit 3 shiploads - others might have storages that are only 0.5% full with 3 shiploads.
> Using relative thresholds instead (like 30% of storage capacity) is generally better, but can also lead to suboptimal situations depending on your fleet, your ore environment, distances and whatnot.

That's why I said that a solid formula would come up during coding :). Edge cases and such is a thing that's always present - and these aiscripts are no easy thing to test. Generally better is, of course, true. You have picked a few words, but you seem to flush all the others from context. Please, read this:

> Probably something exponential on the stock level: effectively zero importance above that threshold, balanced importance around ~3 ship loads, big importance when the warehouse is really starving.

It means "something". ~ means around. Around means I am not sure, I haven't yet written the exact value. And I wont. I have zero intent to spend more than a few minutes on deciding what to choose for the exact value for that. Because I still think that mining as a fallback solves this easily - there's no such problem in that approach.

> I'll gladly discuss this with you.

Discussion is good, but what else can I provide for "mining as a fallback really works good - in my opinion"? Except for "it's my opinion", of course :). Analyzis? I can, but I've already provided some, it's already in the PR header - "Solid/liquid wares (ore, silicon, ice, hydrogen, methane, helium) can only be transported by mining ships, so warehouse networks that distribute raw resources are necessarily crewed by miners. Those ships idle a lot when the network is saturated — and unlike container traders they can do something genuinely useful while idle: mine the very wares their home warehouse stocks." - I still think that mining as a fallback fits well - mod's main activity and all the previous flow stays same, this feature comes into play only when scheduler cannot find a job for some fleet. "Flow stays same" is a crucial thing, obviously. Well, adding mining as a job with another configurable priority sounds well! Sorry, I have lost my chain of thought here :). It's a pure philosophy after that. I am good at philosophy, but sorry again - I do not want to do it right here, right now.
As I said earlier - it's ok to try to reimplement this in another form. I have started already and it does not look good enough for me. But I will try anyway.

I will just add one more commit - with a stop watcher for mining that takes too long, with priority and stuff, and so on - if you wont like the exact calculated values you'll have an opportunity to change it however you like - it's one obvious spot to calculate that importance.

File: md/warehousefleets.xml [hunk 1]
```diff
@@ -245,6 +268,110 @@
...
+                    <do_if value="$stagnantChecks ge 2">
+                      <debug_to_file chance="100" name="$ship.idcode" directory="'WarehouseFleets'"
+                        text="'  MiningOpWatcher: recalling miner (cargo=' + $cargoNow + '/' + $delivery.$amount + ', tripTime=' + (player.age - $delivery.$timeScheduled) + 's)'" output="false" append="true" />
+                      <!-- Cancelling the MiningCollect leaf lets the parent order end normally -->
+                      <do_if value="@$ship.order.id == 'MiningCollect'">
+                        <cancel_order order="$ship.order" />
+                      </do_if>
+                      <do_else>
+                        <cancel_order order="$mineOrder" />
+                      </do_else>
```
Comment (mbleichner) [inline md/warehousefleets.xml:413]: Are you sure this is working correctly?

If I cancel the order manually, the WarehousefleetMining just adds another one. Also I had a ship mining in a sector for like 10 minutes, barely making progress and it didn't get recalled.
